docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,376 @@
+- A provider must register signers of:
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-<provider>`
+    - `cluster-x-k8s-io/kubelet-serving-<provider>`
+- A secure provider MUST implement a secure attestation mechanism, based upon PEM-encoded blocks within the Certificate Signing Request.
+- Where a secure provider’s attestation mechanism does not include a nonce or timestamp to protect against replay attacks, the mechanism MUST implement a secondary time-limited attestation (e.g. AWS Instance Identity document + AWS HMACv4 signature).
+- A provider’s signer MUST run on the management cluster.
+
+##### TPM based providers
+- A TPM provider MUST use the following certificate signers
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-tpm`
+    - `cluster-x-k8s-io/kubelet-serving-tpm`
+- A TPM provider MUST annotate new CSRs as follows:
+    - Key: cluster-x-k8s-io/tpm-endorsement-key
+    - Value: Platform-specific endorsement key (e.g., retrieved from GCP Shielded VM API or VMware vCenter).
+
+#### Kubeadm
+Since this proposal essentially takes over part of the node registration process from kubeadm, we will require the following changes:
+- kubeadm COULD allow opt-out of kubeadm setting up ClusterRoleBindings between the system:nodes group and the `system:certificates.k8s.io:certificatesigningrequests:selfnodeclient` permission, so that certificate renewals must go through re-attestation.
+- Kubeadm COULD allow opt-out of kubeadm setting up `kubeadm:node-autoapprove-bootstrap` cluster role binding.
+This is deferred to a future Kubeadm design and release, and for this proposal, we will add fields to KubeadmControlPlane to remove these node groups and bindings post control plane initialisation.
+
+The idea is to rely the [exec mechanism](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#configuration) of kubeconfigs with local cache directory, when kubelet wants to talk to the apiserver it will call on the kubelet authenticator to get a client certificate.
+
+#### Changes to Cluster
+
+```
+spec:
+  security:
+    kubeletAuthentication: true
+```
+
+#### Changes to KCP
+
+The cluster field if set will be read by KCP and remove the `kubeadm:node-autoapprove-bootstrap` cluster role binding.
+
+A new role binding will be set up to ensure CSRs can still be submitted successfully.
+
+#### Changes to CAPBK
+
+If the kubeletAuthentication field is set for the cluster, CABPK will
+default `--rotate-server-certificates` on NodeRegistrationOptions.ExtraArgs for the Kubeadm
+configuration. If KubeletConfiguration is supported within Cluster API v1alpha4, we
+will opt to set ServerTLSBootstrap on KubeletConfiguration instead.
+
+CABPK will also update the runcmds for cloud-init / Ignition such that the Kubelet kubeconfig
+has the kubelet authentication plugin in exec. This is subject to experimentation,
+and may not be required, however, if this the kubeconfig is built into image builder.
```

Comment:
I think the last part of the sentence doesn't make sense

Reply:
Cleaned text up.